mnemosyne/core/beam.py

```
@@ -3780,14 +3780,28 @@ def get_context(self, limit: int = 10) -> List[Dict]:
+        # unprotected, a "database is locked" here (e.g. while a consolidation
+        # pass is writing) abandons the connection inside an open, stale
+        # transaction -- and because the connection is a long-lived thread
+        # local, EVERY later write on this thread then fails "database is
+        # locked" instantly (stale-snapshot upgrade; the busy handler is not
+        # consulted) until something resets it.
+        try:
+            cursor.execute("BEGIN TRANSACTION")
+            for ts, ids in updates.items():
+                placeholders = ",".join("?" for _ in ids)
+                cursor.execute(
+                    f"UPDATE working_memory SET recall_count = recall_count + 1, last_recalled = ? WHERE id IN ({placeholders})",
```

Ruff (0.15.20)

[error] Possible SQL injection vector through string-based query construction

(S608)